shell/app-shell/elements/shell-ui.js

```diff
@@ -210,6 +211,75 @@ class ShellUi extends Xen.Debug(Xen.Base, log) {
   _onShare(e, share) {
     this._fire('share', share);
   }
+  _onMicStart() {
+    this._setState({micHidden: false});
+  }
+  _onMicEnd(e) {
+    this._setState({micHidden: true});
+    const finalTranscript = e.detail.toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+    log(`voice trigger: [${finalTranscript}]`);
+    if (!finalTranscript) {
+      return;
+    }
+    // see if `node` is a better match for `search` than `match`
+    const findMatch = (node, trigger, search, match) => {
+      log(`node "${node.localName}" has trigger "${trigger}"`);
+      if (trigger.includes(search)) {
+        const diff = trigger.length - search.length;
+        if (!match || match.diff > diff) {
+          log(`matched "${trigger}"`);
+          match = {diff, node};
+        }
+      }
+      return match;
+    };
+    // find all nodes with `trigger` attribute
+    let nodes = document.querySelectorAll('[trigger]');
+    nodes = Array.from(nodes).concat(Array.from(this.host.querySelectorAll('[trigger]')));
+    // find the node that contains `finalTranscript` with the least number of non-matching characters
+    let match;
+    for (const node of nodes) {
+      const trigger = node.getAttribute('trigger').toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+      match = findMatch(node, trigger, finalTranscript, match);
+    }
+    // if we matched, value is the transcript
+    if (match) {
+      match.value = finalTranscript;
+    }
+    // if not, look for a prefix-match
+    else {
+      for (const node of nodes) {
+        const trigger = node.getAttribute('trigger').toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+        if (finalTranscript.startsWith(trigger)) {
+          match = {
+            node: node,
+            value: finalTranscript.slice(trigger.length).trim(),
+          };
+          break;
+        }
+      }
+    }
+    // if there is a matching node, install value and click it
+    if (match) {
+      match.node.value = match.value;
```

Comment:
Seems worth logging that we did this (speech-triggered click, and on what).

Reply:
In general, I try to be careful about adding things to the log stream since it's already crowded. I'm inclined to think we aren't going to need logging to debug problems here, and in any case I suspect we can infer what happened from the logging in `findMatch`.
If we find ourselves asking "what happened?" then we definitely should add more logging.